Core/include/Acts/Utilities/TrackHelpers.hpp

```diff
@@ -0,0 +1,339 @@
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Tolerance.hpp"
+#include "Acts/EventData/MultiTrajectoryHelpers.hpp"
+#include "Acts/EventData/TrackStateType.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/Propagator/StandardAborters.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "Acts/TrackFitting/GainMatrixSmoother.hpp"
+#include "Acts/Utilities/Logger.hpp"
+#include "Acts/Utilities/Result.hpp"
+
+#include <utility>
+
+namespace Acts {
+
+enum class TrackExtrapolationStrategy {
+  /// Use the first track state to reach target surface
+  first,
+  /// Use the last track state to reach target surface
+  last,
+  /// Use the first or last track state to reach target surface depending on the
+  /// distance
+  firstOrLast,
+};
+
+enum class TrackExtrapolationError {
+  CompatibleTrackStateNotFound = 1,
+  ReferenceSurfaceUnreachable = 2,
+};
+
+std::error_code make_error_code(TrackExtrapolationError e);
+
+/// @brief Smooth a track using the gain matrix smoother
+///
+/// @tparam track_container_t The track container type
+///
+/// @param geoContext The geometry context
+/// @param trackContainer The track container
+/// @param track The track to smooth
+/// @param logger The logger
+///
+/// @return The result of the smoothing
+template <typename track_container_t>
+Result<void> smoothTrack(
+    const GeometryContext &geoContext, track_container_t &trackContainer,
+    typename track_container_t::TrackProxy &track,
+    const Logger &logger = *getDefaultLogger("TrackSmoother", Logging::INFO)) {
+  Acts::GainMatrixSmoother smoother;
+
+  auto smoothingResult =
+      smoother(geoContext, trackContainer.trackStateContainer(),
+               track.tipIndex(), logger);
+
+  if (!smoothingResult.ok()) {
+    ACTS_ERROR("Smoothing track " << track.index() << " failed with error "
+                                  << smoothingResult.error());
+    return smoothingResult.error();
+  }
+
+  return Result<void>::success();
+}
+
+/// @brief Smooth tracks using the gain matrix smoother
+///
+/// @tparam track_container_t The track container type
+///
+/// @param geoContext The geometry context
+/// @param trackContainer The track container
+/// @param logger The logger
+///
+/// @return The result of the smoothing
+template <typename track_container_t>
+Result<void> smoothTracks(
+    const GeometryContext &geoContext, const track_container_t &trackContainer,
+    const Logger &logger = *getDefaultLogger("TrackSmoother", Logging::INFO)) {
+  Result<void> result = Result<void>::success();
+
+  for (const auto &track : trackContainer) {
+    auto smoothingResult =
+        smoothTrack(geoContext, trackContainer, track, logger);
+
+    if (!smoothingResult.ok()) {
+      result = smoothingResult.error();
```

Comment:
Theoretically, there could be multiple errors I guess, do we think this could be a problem due to only the last error being visible?

Reply:
Yes I guess this is the tradeoff for that interface. I think it is "okay" as we still propagate at least one error and the user will decide what to do with it.

Alternatively we could use error counters or a vector of error results?

Reply:
Would it be more useful to keep only the *first* error?

Reply:
good question - do you see a benefit in using the first instead of the last?

Reply:
if it were complaining about the same track, then the first error (eg. filter failure) might lead to the second (eg. smoothing failure). If it is separate seeds/tracks, then I think it is all the same.